File: test_mqtt.js
#!/usr/bin/env node
/**
 * MQTT Connectivity & State Machine Test Script
 * Tests Opta1 (energy master) and Opta2 (hottub controller)
 * Simulates energy meter values and monitors system responses
 */

const mqtt = require('mqtt');

// Configuration
const BROKER_IP = '192.168.0.10';
const BROKER_PORT = 1883;
const BROKER_URL = `mqtt://${BROKER_IP}:${BROKER_PORT}`;

// Energy meter topics (Opta's listen to these)
const TOPICS_METER = {
  CH1: 'b0b21c913c34/PUB/CH1',      // fase 1 export
  CH10: 'b0b21c913c34/PUB/CH10',    // fase 1 import
  CH13: 'b0b21c913c34/PUB/CH13',    // totaal export
  CH14: 'b0b21c913c34/PUB/CH14',    // totaal import
};

// Opta1 status topics (we subscribe to these)
const TOPICS_STATUS = {
  surplus_f1: 'opta1/status/surplus_fase1_w',
  surplus_totaal: 'opta1/status/surplus_totaal_w',
  priority: 'opta1/status/active_priority',
  mqtt_valid: 'opta1/status/mqtt_valid',
  mqtt_lastseen: 'opta1/status/mqtt_last_seen',
  wp_active: 'opta1/status/wp_active',
  element_active: 'opta1/status/element_active',
  hottub_perm: 'opta1/status/hottub_permission',
  alarms: 'opta1/status/alarms',
  heartbeat: 'opta1/device/heartbeat',
};

// Opta2 status topics
const TOPICS_OPTA2 = {
  hottub_status: 'opta2/status/hottub_status',
  opta2_alarms: 'opta2/status/alarms',
};

// Test Scenarios
const SCENARIOS = {
  idle: {
    description: 'Geen surplus - alle lasten uit',
    CH1: 100.0,
    CH10: 100.0,    // fase1 surplus = 0W
    CH13: 150.0,
    CH14: 150.0,    // totaal surplus = 0W
    expected_state: 'IDLE',
    duration: 5,
  },
  wp_active: {
    description: 'WP zou actief moeten zijn (fase1 overschot, temp laag)',
    CH1: 1500.0,
    CH10: 100.0,    // fase1 surplus = 1400W
    CH13: 2000.0,
    CH14: 100.0,    // totaal surplus = 1900W
    expected_state: 'WP_BOILER',
    duration: 5,
  },
  drop_to_idle: {
    description: 'Surplus weg - terug naar IDLE',
    CH1: 100.0,
    CH10: 100.0,    // fase1 surplus = 0W
    CH13: 150.0,
    CH14: 150.0,    // totaal surplus = 0W
    expected_state: 'IDLE',
    duration: 5,
  },
};

// Global state
let receivedData = {};
let client = null;
let testActive = true;

// MQTT event handlers
function setupMqtt() {
  return new Promise((resolve, reject) => {
    console.log(`\nConnecting to MQTT broker at ${BROKER_URL}...\n`);
    
    client = mqtt.connect(BROKER_URL, {
      reconnectPeriod: 1000,
      connectTimeout: 5000,
      clientId: 'zonnestroom_test_client_' + Date.now(),
    });

    client.on('connect', () => {
      console.log('✓ MQTT Connected\n');
      
      // Subscribe to all status topics
      const allTopics = [...Object.values(TOPICS_STATUS), ...Object.values(TOPICS_OPTA2)];
      console.log('Subscribing to status topics:');
      allTopics.forEach(topic => {
        client.subscribe(topic, (err) => {
          if (!err) console.log(`  → ${topic}`);
        });
      });
      
      console.log('');
      resolve();
    });

    client.on('message', (topic, message) => {
      const payload = message.toString('utf-8');
      const timestamp = new Date().toLocaleTimeString();
      
      receivedData[topic] = {
        value: payload,
        timestamp: timestamp,
      };
      
      console.log(`\n📨 [${timestamp}] RX: ${topic}`);
      console.log(`   → Value: ${payload}`);
    });

    client.on('error', (err) => {
      console.error(`✗ MQTT Error: ${err.message}`);
      reject(err);
    });

    client.on('disconnect', () => {
      console.log('\n✓ Disconnected from MQTT broker');
    });

    setTimeout(() => {
      if (!client.connected) {
        reject(new Error('Failed to connect within timeout'));
      }
    }, 6000);
  });
}

// Publish meter values
function publishMeterValues(ch1, ch10, ch13, ch14) {
  console.log('\n📤 Publishing meter values:');
  
  const values = { CH1: ch1, CH10: ch10, CH13: ch13, CH14: ch14 };
  
  Object.entries(values).forEach(([channel, value]) => {
    const topic = TOPICS_METER[channel];
    const payload = JSON.stringify({ P: Math.round(value) });
    client.publish(topic, payload);
    console.log(`   ${channel.padEnd(5)} → ${Math.round(value).toString().padStart(6)}W`);
  });
}

// Run a test scenario
async function runTestScenario(name, scenario) {
  console.log('\n' + '='.repeat(70));
  console.log('SCENARIO: ' + name.toUpperCase());
  console.log('='.repeat(70));
  console.log(`Description: ${scenario.description}`);
  console.log(`Expected state: ${scenario.expected_state}`);
  console.log(`Duration: ${scenario.duration}s`);
  
  receivedData = {};
  
  // Publish meter values
  publishMeterValues(scenario.CH1, scenario.CH10, scenario.CH13, scenario.CH14);
  
  // Wait for responses
  console.log(`\nWaiting for responses...`);
  for (let i = scenario.duration; i > 0; i--) {
    await new Promise(resolve => setTimeout(resolve, 1000));
    process.stdout.write(`\r   ... ${i}s remaining`);
  }
  console.log('\n');
  
  // Display collected responses
  console.log('Responses collected during scenario:');
  if (Object.keys(receivedData).length === 0) {
    console.log('   (no responses - check MQTT broker connectivity!)');
  } else {
    Object.entries(receivedData).forEach(([topic, data]) => {
      console.log(`   ${topic}`);
      console.log(`      → ${data.value} (${data.timestamp})`);
    });
  }
}

// Print checklist
function printChecklist() {
  console.log('\n' + '='.repeat(70));
  console.log('COMMISSIONING CHECKLIST');
  console.log('='.repeat(70) + '\n');
  
  const checks = [
    ['MQTT Broker connectivity', 'Connection successful if above'],
    ['Opta1 receives meter values', 'CH1, CH10, CH13, CH14 topics'],
    ['Opta1 calculates surplus', 'surplus_fase1_w, surplus_totaal_w'],
    ['Opta1 state machine works', 'priority state transitions'],
    ['Opta1 publishes status', 'wp_active, element_active, etc.'],
    ['Opta2 heartbeat detected', 'opta1/device/heartbeat toggling'],
    ['Opta2 hottub control works', 'opta2/status/hottub_status'],
    ['Fail-safe interlocks active', 'Check alarms when expected'],
  ];
  
  checks.forEach(([name, desc]) => {
    console.log(`  ☐ ${name}`);
    console.log(`      ${desc}\n`);
  });
}

// Main execution
async function main() {
  try {
    console.log('\n' + '='.repeat(70));
    console.log('ZONNESTROOM DUAL-OPTA MQTT TEST');
    console.log('='.repeat(70));
    
    // Setup MQTT
    await setupMqtt();
    
    // Give broker time to confirm subscriptions
    await new Promise(resolve => setTimeout(resolve, 2000));
    
    // Run scenarios
    for (const [name, scenario] of Object.entries(SCENARIOS)) {
      await runTestScenario(name, scenario);
      await new Promise(resolve => setTimeout(resolve, 1000));
    }
    
    // Print checklist
    printChecklist();
    
    // Monitor for 30 seconds
    console.log('📊 Monitoring MQTT topics for 30 more seconds...\n');
    await new Promise(resolve => setTimeout(resolve, 30000));
    
    console.log('\n✓ Test completed successfully!');
    
  } catch (err) {
    console.error('\n✗ Error:', err.message);
    console.error('\nTroubleshooting:');
    console.error('  • Is MQTT broker running at 192.168.0.10:1883?');
    console.error('  • Are both Opta boards connected via Ethernet?');
    console.error('  • Check network connectivity: ping 192.168.0.10');
    process.exit(1);
  } finally {
    testActive = false;
    if (client) {
      client.end();
    }
  }
}

// Run if called directly
if (require.main === module) {
  main();
}

module.exports = { setupMqtt, publishMeterValues, runTestScenario };

File: scripts/test_mqtt.py
#!/usr/bin/env python3
"""
MQTT Connectivity & State Machine Test Script
Tests Opta1 (energy master) and Opta2 (hottub controller)
Simulates energy meter values and monitors system responses
"""

import paho.mqtt.client as mqtt
import json
import time
import sys
from typing import Dict, List
from datetime import datetime

# ─── Configuration ─────────────────────────────────────────────────────────
BROKER_IP = "192.168.0.10"
BROKER_PORT = 1883

# Energy meter topics (Opta's listen to these)
TOPICS_METER = {
    "CH1": "b0b21c913c34/PUB/CH1",      # fase 1 export
    "CH10": "b0b21c913c34/PUB/CH10",    # fase 1 import
    "CH13": "b0b21c913c34/PUB/CH13",    # totaal export
    "CH14": "b0b21c913c34/PUB/CH14",    # totaal import
}

# Opta1 status topics (we subscribe to these)
TOPICS_STATUS = {
    "surplus_f1": "opta1/status/surplus_fase1_w",
    "surplus_totaal": "opta1/status/surplus_totaal_w",
    "priority": "opta1/status/active_priority",
    "mqtt_valid": "opta1/status/mqtt_valid",
    "mqtt_lastseen": "opta1/status/mqtt_last_seen",
    "wp_active": "opta1/status/wp_active",
    "element_active": "opta1/status/element_active",
    "hottub_perm": "opta1/status/hottub_permission",
    "alarms": "opta1/status/alarms",
    "heartbeat": "opta1/device/heartbeat",
}

# Opta2 status topics
TOPICS_OPTA2 = {
    "hottub_status": "opta2/status/hottub_status",
    "alarms": "opta2/status/alarms",
}

# ─── Test Scenarios ────────────────────────────────────────────────────────
SCENARIOS = {
    "idle": {
        "description": "Geen surplus - alle lasten uit",
        "CH1": 100.0,
        "CH10": 100.0,    # fase1 surplus = 0W
        "CH13": 150.0,
        "CH14": 150.0,    # totaal surplus = 0W
        "expected_state": "IDLE",
        "duration": 5,
    },
    "wp_active": {
        "description": "WP zou actief moeten zijn (fase1 overschot, temp laag)",
        "CH1": 1500.0,
        "CH10": 100.0,    # fase1 surplus = 1400W
        "CH13": 2000.0,
        "CH14": 100.0,    # totaal surplus = 1900W
        "expected_state": "WP_BOILER",
        "duration": 5,
    },
    "drop_to_idle": {
        "description": "Surplus weg - terug naar IDLE",
        "CH1": 100.0,
        "CH10": 100.0,    # fase1 surplus = 0W
        "CH13": 150.0,
        "CH14": 150.0,    # totaal surplus = 0W
        "expected_state": "IDLE",
        "duration": 5,
    },
}

# ─── Global State ──────────────────────────────────────────────────────────
received_data = {}
test_active = True
client = None


# ─── MQTT Callbacks ────────────────────────────────────────────────────────
def on_connect(client, userdata, flags, rc):
    """Called when connected to broker"""
    if rc == 0:
        print(f"\n✓ MQTT Connected (code {rc})\n")
        
        # Subscribe to all status topics
        for topic in TOPICS_STATUS.values():
            client.subscribe(topic)
            print(f"  → Subscribed to: {topic}")
        
        for topic in TOPICS_OPTA2.values():
            client.subscribe(topic)
            print(f"  → Subscribed to: {topic}")
            
    else:
        print(f"✗ Connection failed with code {rc}")
        sys.exit(1)


def on_message(client, userdata, msg):
    """Called when a message is received"""
    global received_data
    
    # Store received value
    topic = msg.topic
    payload = msg.payload.decode('utf-8', errors='ignore')
    
    received_data[topic] = {
        'value': payload,
        'timestamp': datetime.now(),
    }
    
    # Pretty print incoming data
    print(f"\n📨 RX: {topic}")
    print(f"   → Value: {payload}")


def on_disconnect(client, userdata, rc):
    """Called when disconnected"""
    if rc != 0:
        print(f"\n✗ Unexpected MQTT disconnection: code {rc}")
    else:
        print(f"\n✓ Disconnected from MQTT broker")


# ─── Test Functions ────────────────────────────────────────────────────────
def publish_meter_values(ch1: float, ch10: float, ch13: float, ch14: float):
    """Publish simulated energy meter values"""
    values = {
        "CH1": ch1,
        "CH10": ch10,
        "CH13": ch13,
        "CH14": ch14,
    }
    
    print(f"\n📤 Publishing meter values:")
    for channel, value in values.items():
        topic = TOPICS_METER[channel]
        payload = f'{{"P":{int(value)}}}'
        client.publish(topic, payload, qos=1, retain=False)
        print(f"   {channel:5s} → {int(value):6d}W ({topic})")
    
    # Publish timestamp so Opta knows when last meter update was
    client.publish(f"{TOPICS_METER['CH1'].rsplit('/', 1)[0]}/timestamp", 
                   str(int(time.time())), qos=1, retain=False)


def run_test_scenario(scenario_name: str, scenario: Dict):
    """Execute a test scenario"""
    global received_data
    
    print(f"\n{'='*70}")
    print(f"SCENARIO: {scenario_name.upper()}")
    print(f"{'='*70}")
    print(f"Description: {scenario['description']}")
    print(f"Expected state: {scenario['expected_state']}")
    print(f"Duration: {scenario['duration']}s\n")
    
    # Publish meter values
    publish_meter_values(
        scenario["CH1"],
        scenario["CH10"],
        scenario["CH13"],
        scenario["CH14"]
    )
    
    # Wait for responses
    remaining = scenario['duration']
    while remaining > 0 and test_active:
        time.sleep(1)
        remaining -= 1
        print(f"   ... waiting ({remaining}s remaining)")
    
    # Display collected responses
    print(f"\nResponses collected during scenario:")
    if not received_data:
        print("   (no responses yet - check MQTT broker connectivity)")
    else:
        for topic, data in sorted(received_data.items()):
            print(f"   {topic}")
            print(f"      → {data['value']}")
    
    received_data.clear()


def run_connectivity_test():
    """Test basic MQTT connectivity"""
    print(f"\n{'='*70}")
    print("CONNECTIVITY TEST")
    print(f"{'='*70}")
    print(f"Broker: {BROKER_IP}:{BROKER_PORT}\n")
    
    # This will be tested in on_connect callback
    print("Waiting for connection...")


def print_summary():
    """Print commissioning summary"""
    print(f"\n{'='*70}")
    print("COMMISSIONING CHECKLIST")
    print(f"{'='*70}\n")
    
    checks = [
        ("MQTT Broker connectivity", "Check connection message above"),
        ("Opta1 receives meter values", "CH1, CH10, CH13, CH14"),
        ("Opta1 calculates surplus", "surplus_fase1_w, surplus_totaal_w"),
        ("Opta1 state machine works", "priority state transitions"),
        ("Opta1 publishes status", "wp_active, element_active, etc."),
        ("Opta2 heartbeat detected", "opta1/device/heartbeat toggling"),
        ("Opta2 hottub control works", "opta2/status/hottub_status"),
        ("Fail-safe interlocks active", "Check alarms when expected"),
    ]
    
    for name, description in checks:
        print(f"  ☐ {name}")
        print(f"      {description}\n")


# ─── Main ──────────────────────────────────────────────────────────────────
def main():
    global client, test_active
    
    try:
        # Connect to MQTT broker
        client = mqtt.Client(client_id="zonnestroom_test_client")
        client.on_connect = on_connect
        client.on_message = on_message
        client.on_disconnect = on_disconnect
        
        print(f"\n{'='*70}")
        print("ZONNESTROOM DUAL-OPTA MQTT TEST")
        print(f"{'='*70}")
        print(f"\nConnecting to MQTT broker at {BROKER_IP}:{BROKER_PORT}...")
        
        client.connect(BROKER_IP, BROKER_PORT, keepalive=60)
        client.loop_start()
        
        # Wait for connection
        time.sleep(2)
        
        if not client.is_connected():
            print("✗ Failed to connect to MQTT broker")
            print("   Ensure:")
            print("   1. Broker is running at 192.168.0.10:1883")
            print("   2. Both Opta boards are connected via Ethernet")
            print("   3. Network connectivity is configured")
            sys.exit(1)
        
        # Run connectivity test first
        run_connectivity_test()
        time.sleep(2)
        
        # Run test scenarios
        for scenario_name, scenario in SCENARIOS.items():
            run_test_scenario(scenario_name, scenario)
            time.sleep(1)
        
        # Print summary
        print_summary()
        
        # Keep running to display real-time updates
        print("\n📊 Monitoring MQTT for 30 seconds (Ctrl+C to exit)...\n")
        time.sleep(30)
        
    except KeyboardInterrupt:
        print("\n\n✓ Test completed by user")
    except Exception as e:
        print(f"\n✗ Error: {e}")
        import traceback
        traceback.print_exc()
    finally:
        test_active = False
        if client:
            client.loop_stop()
            client.disconnect()


if __name__ == "__main__":
    main()
